File: src/api/app_state.ts
import { invoke } from "@tauri-apps/api/core";

export async function appStateGet(key: string): Promise<string | null> {
  return invoke("app_state_get", { key });
}

export async function appStateSet(key: string, value: string): Promise<void> {
  return invoke("app_state_set", { key, value });
}

export async function appStateDelete(key: string): Promise<void> {
  return invoke("app_state_delete", { key });
}

export async function appStateGetAll(): Promise<Record<string, string>> {
  return invoke("app_state_get_all");
}

File: src-tauri/src/commands/app_state.rs
use std::collections::HashMap;

use serde::Serialize;
use tauri::State;

use crate::error::AppResult;
use crate::state::AppState;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AppStateEntry {
    pub key: String,
    pub value: String,
}

#[tauri::command]
pub async fn app_state_get(state: State<'_, AppState>, key: String) -> AppResult<Option<String>> {
    state.usage_db.state_get(&key)
}

#[tauri::command]
pub async fn app_state_set(
    state: State<'_, AppState>,
    key: String,
    value: String,
) -> AppResult<()> {
    state.usage_db.state_set(&key, &value)
}

#[tauri::command]
pub async fn app_state_delete(state: State<'_, AppState>, key: String) -> AppResult<()> {
    state.usage_db.state_delete(&key)
}

#[tauri::command]
pub async fn app_state_get_all(state: State<'_, AppState>) -> AppResult<HashMap<String, String>> {
    let pairs = state.usage_db.state_get_all()?;
    Ok(pairs.into_iter().collect())
}
